File: TrainerApi/Services/TrainerService.cs
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;

namespace TrainerApi.Services;


public class TrainerService : TrainerApi.TrainerService.TrainerServiceBase{

    private readonly ITtainerRepository ttainerRepository
    public 
    public override async Task<TrainerResponse> GetTrainer(TrainerByIdRequest request, ServerCallContext context)
    {
        return new TrainerResponse{
            Id = Guid.NewGuid().ToString(),
            Name = "Pascual",
            Age = 99,
            Birthdate = Timestamp.FromDateTime(DateTime.UtcNow),
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow),
            Medals    = {
            new Medals{ Region = "MX", Type = MedalsType.Gold },
            new Medals{ Region = "JP", Type = MedalsType.Silver }
        }
    };
}
public override async Task<TrainerResponse> CreateTrainer(TrainerRequest request, ServerCallContext context)
{
   var trainer = await _trainerRepository.GetByIdAsync(request.Id, context.CancellationToken)
   if (trainer is null){
    throw new RcpException(new Status(new Status(StatusCode.NotFound, "Trainer not found")));
   }
   return trainer.ToResponse();
}
}

File: TrainerGrpcService/Services/trainer_service.py
from concurrent import futures
import grpc
import uuid
from google.protobuf.timestamp_pb2 import Timestamp
import datetime

import trainer_pb2
import trainer_pb2_grpc

class TrainerService(trainer_pb2_grpc.TrainerServiceServicer):
    def GetTrainer(self, request, context):
        now = datetime.datetime.utcnow()
        birthdate = Timestamp()
        birthdate.FromDatetime(now)

        created_at = Timestamp()
        created_at.FromDatetime(now)

        return trainer_pb2.TrainerResponse(
            id=str(uuid.uuid4()),
            name="Pascual",
            age=99,
            birthdate=birthdate,
            createdAt=created_at,
            medals=[
                trainer_pb2.Medals(region="MX", type=trainer_pb2.MedalsType.Gold),
                trainer_pb2.Medals(region="JP", type=trainer_pb2.MedalsType.Silver)
            ]
        )

def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    trainer_pb2_grpc.add_TrainerServiceServicer_to_server(TrainerService(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    print("TrainerService gRPC server running on port 50051...")
    server.wait_for_termination()

if __name__ == "__main__":
    serve()
